src/rdpc/findMatchedTNPairs.ts

```diff
@@ -0,0 +1,99 @@
+import _ from "lodash";
+import {
+  DonorData,
+  SamplePair,
+  StringMap,
+  TumourNormalDesignationValue,
+} from "rdpc/types";
+
+// iterates over donor's DonorData and finds all the matched
+// tumour/normal sample pairs. Matched samples must have the same
+// experiment strategy; tumour sample uses
+// matchedNormalSubmitterSampleId field as a foreign key
+// to look for the matched normal sample.
+export const findMatchedTNPairs = (
+  map: StringMap<DonorData>
+): StringMap<SamplePair[]> => {
+  const donorsWithSamplePairs: StringMap<SamplePair[]> = {};
+
+  Object.entries(map).forEach(([donorId, donorData]) => {
+    const allSamplePairs: SamplePair[] = [];
+    donorData.samplePairs.forEach((currentSample) => {
+      if (currentSample.tumourSample) {
+        const matchedNormalSample = donorData.samplePairs.filter(
+          (sample) =>
+            sample.normalSample &&
+            sample.normalSample.tumourNormalDesignation ===
+              TumourNormalDesignationValue.Normal &&
+            !sample.tumourSample &&
+            sample.normalSample.experimentStrategy ===
+              currentSample.tumourSample?.experimentStrategy &&
+            sample.normalSample.submitterSampleId ===
+              currentSample.tumourSample?.matchedNormalSubmitterSampleId
+        );
+
+        matchedNormalSample.forEach((matched) => {
+          const samplePair = {
+            normalSample: matched.normalSample,
+            tumourSample: currentSample.tumourSample,
+            firstPublishedAt: determinSamplePairFirstPublishedDate(
+              matched,
+              currentSample
+            ),
+          };
+          allSamplePairs.push(samplePair);
+        });
+      }
+    });
+
+    if (donorsWithSamplePairs[donorId]) {
+      donorsWithSamplePairs[donorId] = [
+        ...donorsWithSamplePairs[donorId],
+        ...allSamplePairs,
+      ];
+    } else {
+      donorsWithSamplePairs[donorId] = allSamplePairs;
+    }
+  });
+
+  return donorsWithSamplePairs;
+};
+
+// returns the most recent first published date as the earliest available sample pair published date
+const determinSamplePairFirstPublishedDate = (
```

Comment:
See other comment, let's just put the content of this method inline since it is only used in one place and we don't need the complexity of the error handling.